### DISCUSSION TITLE
Testing with auth

### DISCUSSION THREAD

**Discussion by Minivera** [encoredev/encore#76] [Help] [answered]
Quick question, I have an endpoint protected by an authHandler in which I extract the user data with `userData := auth.Data().(*identity.UserData)`, where `*identity.UserData` is a struct that contains some user info. This endpoint uses the user data to fetch a list of collections owned by the authenticated user (See code bellow). I'd like to test that endpoint, but I am not entire sure how to make sure that the auth data is available.

```go
//encore:api auth
func ListCollections(ctx context.Context) (*ListCollectionsResponse, error) {
	userData := auth.Data().(*identity.UserData)

	collections, err := models.ListCollections(ctx, userData.ID)
	if err != nil {
		log.WithError(err).Warning("Could not fetch collections for this user")
		return nil, &errs.Error{
			Code:    errs.Internal,
			Message: "Could not fetch collections",
		}
	}

	return &ListCollectionsResponse{
		Collections: convert.CollectionModelsToPayloads(collections),
	}, nil
}
```

I took a look at the runtime code from encode (Which is exposed by the `encode.dev` package) and I thought I could try `runtime.BeginRequest` to potentially trick encore into thinking there is some auth data, but that function is not exposed by `encore.dev` (And that probably would have been very hacky). Would you happen to have a suggestion on how to test endpoints that use an auth handler?

**Comment by eandre** [comment, marked as answer]
I've thought about this as well and we should definitely do something to make this easier to test (probably add the ability to inject auth data into an API call by wrapping the context, or something like that).

Until then, I suggest taking the business logic part and extracting it as a separate function that takes the auth data as a parameter:

```go
//encore:api auth
func ListCollections(ctx context.Context) (*ListCollectionsResponse, error) {
	userData := auth.Data().(*identity.UserData)
	return listCollections(ctx, userData)
}

func listCollections(ctx context.Context, userData *identity.UserData) (*ListCollectionsResponse, error) {
	collections, err := models.ListCollections(ctx, userData.ID)
	if err != nil {
		log.WithError(err).Warning("Could not fetch collections for this user")
		return nil, &errs.Error{
			Code:    errs.Internal,
			Message: "Could not fetch collections",
		}
	}

	return &ListCollectionsResponse{
		Collections: convert.CollectionModelsToPayloads(collections),
	}, nil
}
```

And then write tests against `listCollections` instead.

**Reply by Minivera** [reply]
Will do that, thanks! It will increase the size of the code and reduce test coverage a little, but that's not that big of a problem. I can revisit these tests when encore has a way of injecting auth data and increase coverage for free.

Pure curiosity: did you consider having the auth data and other globals (like the SQL conn) live in the context? I'd be really interested in learning the history behind the `encore.dev` package and its design.
